# **1. Import Library**

Pada tahap ini, Anda perlu mengimpor beberapa pustaka (library) Python yang dibutuhkan untuk analisis data dan pembangunan model machine learning.

In [37]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report
from sklearn.ensemble import RandomForestClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import mean_squared_error

# **2. Memuat Dataset dari Hasil Clustering**

Memuat dataset hasil clustering dari file CSV ke dalam variabel DataFrame.

In [38]:
train_df = pd.read_csv("clustered_co2.csv")
train_df.sample(5)

Unnamed: 0       Make        Model         Vehicle Class  \
3643        3892       JEEP      COMPASS           SUV - SMALL   
2050        2253       AUDI   S5 QUATTRO            SUBCOMPACT   
1274        1298  CHEVROLET     COLORADO  PICKUP TRUCK - SMALL   
4280        4635   CHRYSLER      300 FFV             FULL-SIZE   
2096        2303        BMW  535i xDRIVE              MID-SIZE   

      Engine Size(L)  Cylinders Transmission Fuel Type  \
3643             2.4          4           A6         X   
2050             3.0          6           M6         Z   
1274             2.5          4           A6         X   
4280             3.6          6           A8         X   
2096             3.0          6          AS8         Z   

      Fuel Consumption City (L/100 km)  Fuel Consumption Hwy (L/100 km)  \
3643                              11.3                              8.7   
2050                              13.8                              9.0   
1274                              11.9                              8.8   
4280                              12.4                              7.8   
2096                              11.9                              8.1   

      Fuel Consumption Comb (L/100 km)  Fuel Consumption Comb (mpg)  \
3643                              10.1                           28   
2050                              11.7                           24   
1274                              10.5                           27   
4280                              10.3                           27   
2096                              10.2                           28   

      CO2 Emissions(g/km)  Outlier  cluster  
3643                  238        1        0  
2050                  273        1        1  
1274                  242        1        0  
4280                  242        1        0  
2096                  238        1        1

# **3. Data Splitting**

Tahap Data Splitting bertujuan untuk memisahkan dataset menjadi dua bagian: data latih (training set) dan data uji (test set).

In [39]:
categorical_features = train_df.select_dtypes(include=['object']).columns
X = train_df.drop(columns=['cluster', *categorical_features])
y = train_df['cluster']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=12)

In [40]:
# menghitung panjang/jumlah data
print("Jumlah data: ",len(X))
# menghitung panjang/jumlah data pada x_test
print("Jumlah data latih: ",len(X_train))
# menghitung panjang/jumlah data pada x_test
print("Jumlah data test: ",len(X_test))

Jumlah data:  6282
Jumlah data latih:  5025
Jumlah data test:  1257


# **4. Membangun Model Klasifikasi**


## **a. Membangun Model Klasifikasi**

Setelah memilih algoritma klasifikasi yang sesuai, langkah selanjutnya adalah melatih model menggunakan data latih.

Berikut adalah rekomendasi tahapannya.
1. Pilih algoritma klasifikasi yang sesuai, seperti Logistic Regression, Decision Tree, Random Forest, atau K-Nearest Neighbors (KNN).
2. Latih model menggunakan data latih.

### random forest

In [41]:
RF_model = RandomForestClassifier()
RF_model.fit(X_train, y_train)

RF_y_train_pred = RF_model.predict(X_train)
RF_y_test_pred = RF_model.predict(X_test)

First, submitter will try to use random forest algorithm

### logistic regression

In [42]:
LR_model = LogisticRegression(max_iter=1000)
LR_model.fit(X_train, y_train)

LR_y_train_pred = LR_model.predict(X_train)
LR_y_test_pred = LR_model.predict(X_test)

/home/mendo/projects/python/ml_co2/venv/lib/python3.10/site-packages/sklearn/linear_model/_logistic.py:469: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


submitter also try logistic regression

### Decission Tree

In [43]:
DT_model = DecisionTreeClassifier()
DT_model.fit(X_train, y_train)

DT_y_train_pred = DT_model.predict(X_train)
DT_y_test_pred = DT_model.predict(X_test)

## **b. Evaluasi Model Klasifikasi**

Berikut adalah **rekomendasi** tahapannya.
1. Lakukan prediksi menggunakan data uji.
2. Hitung metrik evaluasi seperti Accuracy dan F1-Score (Opsional: Precision dan Recall).
3. Buat confusion matrix untuk melihat detail prediksi benar dan salah.

### random forest

In [44]:

print("Training Set Classification Report:")
print(classification_report(y_train, RF_y_train_pred))
print( '-' * 50)
print("Test Set Classification Report:")
print(classification_report(y_test, RF_y_test_pred))

Training Set Classification Report:
              precision    recall  f1-score   support

           0       1.00      1.00      1.00      2544
           1       1.00      1.00      1.00      2213
           2       1.00      1.00      1.00       268

    accuracy                           1.00      5025
   macro avg       1.00      1.00      1.00      5025
weighted avg       1.00      1.00      1.00      5025

--------------------------------------------------
Test Set Classification Report:
              precision    recall  f1-score   support

           0       0.92      0.94      0.93       643
           1       0.92      0.90      0.91       552
           2       1.00      1.00      1.00        62

    accuracy                           0.93      1257
   macro avg       0.95      0.95      0.95      1257
weighted avg       0.93      0.93      0.93      1257



evaluation results for the Random Forest algorithm shows that this model have the average as follow: accuracy of 90%, recall of 95% and f1-score of 95% as well.

### logistic regression

In [45]:
print("Training Set Classification Report:")
print(classification_report(y_train, LR_y_train_pred))
print( '-' * 50)
print("Test Set Classification Report:")
print(classification_report(y_test, LR_y_test_pred))

Training Set Classification Report:
              precision    recall  f1-score   support

           0       0.65      0.73      0.69      2544
           1       0.64      0.56      0.60      2213
           2       1.00      1.00      1.00       268

    accuracy                           0.67      5025
   macro avg       0.76      0.76      0.76      5025
weighted avg       0.67      0.67      0.66      5025

--------------------------------------------------
Test Set Classification Report:
              precision    recall  f1-score   support

           0       0.65      0.71      0.68       643
           1       0.62      0.55      0.58       552
           2       1.00      1.00      1.00        62

    accuracy                           0.65      1257
   macro avg       0.76      0.75      0.75      1257
weighted avg       0.65      0.65      0.65      1257



### Decission Tree

In [46]:
print("Training Set Classification Report:")
print(classification_report(y_train, DT_y_train_pred))
print( '-' * 50)
print("Test Set Classification Report:")
print(classification_report(y_test, DT_y_test_pred))

Training Set Classification Report:
              precision    recall  f1-score   support

           0       1.00      1.00      1.00      2544
           1       1.00      1.00      1.00      2213
           2       1.00      1.00      1.00       268

    accuracy                           1.00      5025
   macro avg       1.00      1.00      1.00      5025
weighted avg       1.00      1.00      1.00      5025

--------------------------------------------------
Test Set Classification Report:
              precision    recall  f1-score   support

           0       0.92      0.91      0.91       643
           1       0.89      0.90      0.90       552
           2       1.00      1.00      1.00        62

    accuracy                           0.91      1257
   macro avg       0.94      0.94      0.94      1257
weighted avg       0.91      0.91      0.91      1257



In [47]:
train_mse = mean_squared_error(y_train, DT_y_train_pred)
test_mse = mean_squared_error(y_test, DT_y_test_pred)

print('Decission tree MSE') 
print(f'Training MSE with Decission Tree: {train_mse}')
print(f'Test MSE: {test_mse}')

Decission tree MSE
Training MSE with Decission Tree: 0.0
Test MSE: 0.09148766905330151


In [48]:
train_mse = mean_squared_error(y_train, RF_y_train_pred)
test_mse = mean_squared_error(y_test, RF_y_test_pred)

print('Random Forest MSE') 
print(f'Training MSE: {train_mse}')
print(f'Test MSE: {test_mse}')

Random Forest MSE
Training MSE: 0.0
Test MSE: 0.07478122513922036


## **c. Tuning Model Klasifikasi (Optional)**

Gunakan GridSearchCV, RandomizedSearchCV, atau metode lainnya untuk mencari kombinasi hyperparameter terbaik

In [49]:
#Type your code here

## **d. Evaluasi Model Klasifikasi setelah Tuning (Optional)**

Berikut adalah rekomendasi tahapannya.
1. Gunakan model dengan hyperparameter terbaik.
2. Hitung ulang metrik evaluasi untuk melihat apakah ada peningkatan performa.

In [50]:
#Type your code here

## **e. Analisis Hasil Evaluasi Model Klasifikasi**

Submitter applied three different algorithm Random Forest, Logistic Regression and Decission Tree. Found that Random Forest and Decission Tree have perfect evaluation score on training set while logistic regression under performed by other.
result shows that random forest have a better evaluation score overall by 95% average for each recall, precission, and f1-score followed by decission tree at 93% each. Which mean that both algorithm are able to truly classify car data to the correct cluster and indicate that the cluster well separated.

Based on the evaluation process there is no tuning needed for Random Forest and Decission Tree while Logistic Regression may need hyperparameter tuning in order to improve overall evaluation score or try different feature.